Initial ansible-test support for collections.

mattclay opened the pull request:
##### SUMMARY

Initial ansible-test support for collections.

This maintains existing support for using ansible-test on the Ansible source while providing the initial refactoring needed to support testing of Ansible collections.

Collections testing support is still an early preview. Many features for testing collections are incomplete, have bugs or are completely missing. Some existing features will change significantly in subsequent PRs.

The primary change in existing behavior is that ansible-test can no longer be invoked from a current working directory outside of the content root being tested.

##### ISSUE TYPE

Feature Pull Request

##### COMPONENT NAME

ansible-test

mattclay committed:
Initial ansible-test support for collections.

mattclay committed:
Include cloud config in delegation payload.

mattclay committed:
Add missing types import and fix `t` shadowing.

mattclay committed:
Fix plugin traceback when config_path not set.

mattclay committed:
Fix encoding issues.

mattclay committed:
Remove unused imports.

mattclay committed:
More encoding fixes.

mattclay committed:
Handle delegation outside exception handler.

mattclay committed:
Inject ssh keys only if not already in place.

mattclay committed:
More defensive approach to getting remote pwd.

mattclay committed:
Add missing string format var.

ci_complete
ci_coverage

mattclay commented:
Full coverage run: https://app.shippable.com/github/ansible/ansible/runs/132593/

The timeouts are unrelated to the changes.

jborean93 commented on test/runner/lib/cli.py:
```diff
@@ -106,8 +113,13 @@ def main():
+            # save delegation args for use once we exit the exception handler
+            delegate_args = (ex.exclude, ex.require, ex.integration_targets)
+
+        if delegate_args:
```
What's the point of this change. I can't see any other usage of `delegate_args` below in this method.

mattclay replied:
This prevents Python 3 from using exception chaining when an unhandled exception is generated in `delegate()`. Since the `Delegate` exception is being used for control flow instead of error reporting, there's little value in including it in the traceback if `delegate()` fails.

jborean93 commented on test/runner/lib/cli.py:
```diff
@@ -614,24 +626,30 @@ def add_environments(parser, tox_version=False, tox_only=False):
-                                  const='.'.join(str(i) for i in sys.version_info[:2]),
-                                  choices=SUPPORTED_PYTHON_VERSIONS,
-                                  help='run from a tox virtualenv: %s' % ', '.join(SUPPORTED_PYTHON_VERSIONS))
+    if data_context().content.is_install:
```
Just verifying `is_install` means that we are running from an Ansible install not a collection? If so it might be better with a different property like `is_ansible` or the inverse, `is_collection`.

mattclay replied:
In various places in the code there are two conditions like this that are used:

- `if data_context().content.is_install:` - When we need to know if we're testing the Ansible source, without caring what kind of content is being tested when we're not.
- `if data_context().content.collection:` - When we need to know that we're testing an Ansible collection.

Currently these two conditions are the inverse of each other. However, if we ever add support for testing pre-collection roles (something I included in an early prototype but left out of this PR), then the difference matters, as we would then end up adding a third condition:

- `if data_context().content.role:` - When we need to know that we're testing an Ansible role.

As for the name `is_install`, at one point I actually named it `is_ansible`, but thought `is_install` was better since it matched the `INSTALL_ROOT` constant. If you still think it makes sense to rename it, I can think of a few options and I'm open to other suggestions:

- `is_ansible`
- `is_source`
- `is_ansible_install`
- `is_ansible_source`

jborean93 replied:
Personally I prefer `is_ansible` or `is_ansible_install` as it makes it more obvious the `is_install` is referring to Ansible's install location and not the collection path. I'm not married to it but I definitely had to look at the code to make sure that is correct.

jborean93 commented on test/runner/lib/coverage_util.py:
```diff
@@ -0,0 +1,89 @@
+
+def generate_collection_coverage_config(args):  # type: (TestConfig) -> str
+    """Generate code coverage configuration for tests."""
+    coverage_config = '''
```
Does this mean that `ansible-test` in Ansible will not use the `.coveragerc` file in the root directory.

mattclay replied:
The `.coveragerc` in the root directory will not be used when testing collections, but will continue to be used when testing Ansible itself. I expect to change this in a future PR so that all coverage configuration is generated. At that point we'll be able to remove the static config file.

jborean93 commented on test/runner/lib/powershell_import_analysis.py:
```diff
@@ -62,7 +82,7 @@ def extract_powershell_module_utils_imports(path, module_utils):
         for line in lines:
             line_number += 1
-            match = re.search(r'(?i)^#\s*requires\s+-module(?:s?)\s*(Ansible\.ModuleUtils\..+)', line)
+            match = re.search(r'(?i)^#\s*(?:requires\s+-module|ansiblerequires\s+-powershell)(?:s?)\s*((?:Ansible|AnsibleCollections)\..+)', line)
```
you have an option `s` after either `-module` or `-powershell`. Only module van be either `module` or `modules`.

https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/powershell/module_manifest.py#L40-L41

mattclay replied:
Good catch, that mistake was not intentional. I'll push a commit to correct that.

jborean93 commented on test/runner/lib/sanity/__init__.py:
```diff
@@ -145,8 +148,13 @@ def collect_code_smell_tests():
     :rtype: tuple[SanityFunc]
     """
     skip_file = 'test/sanity/code-smell/skip.txt'
+    ansible_only_file = os.path.join(INSTALL_ROOT, 'test/sanity/code-smell/ansible-only.txt')
```
What does this file signify?

mattclay replied:
This file lists the code-smell sanity tests which are currently executed only on Ansible and not collections. Some of these may updated to work on collections later.

jborean93 commented on test/runner/lib/sanity/ansible_doc.py:
```diff
@@ -67,19 +75,19 @@ def test(self, args, targets, python_version):
 
         for module in modules:
-            doc_targets['module'].append(module)
+            doc_targets['module'].append(data_context().content.prefix + module)
```
should this be `os.path.join` instead?

mattclay replied:
No, since `data_context().content.prefix` is not a path.

When testing Ansible it is `""` and when testing a collection it is in the form `"{namespace}.{collection}."` so that collection names passed to ansible-doc are fully qualified.

jborean93 commented on test/runner/lib/util.py:
```diff
@@ -617,8 +651,8 @@ def print_message(self, message, color=None, fd=sys.stdout, truncate=False):  #
-        if sys.version_info[0] == 2 and isinstance(message, type(u'')):
-            message = message.encode('utf-8')
+        if sys.version_info[0] == 2:
+            message = to_bytes(message)
```
I've never seen `print(file=...)` before, should we still be printing a byte string or should it be a native string?

mattclay replied:
This change just standardizes on the new `to_bytes` function instead of using `encode()` to do the same thing. I was trying to minimize the amount of encoding related changes in this PR that may actually change behavior, other than to fix issues I was encountering during testing.

I'd like to do more encoding cleanup in a follow up PR, including taking another look at this. I just didn't want to do it all at once.

jborean93 reviewed (commented):
Sorry it took so long, everything seems to be good I just had a few questions/queries about some of the changes.

mattclay committed:
Correct PowerShell require regex.

The module option can be plural but powershell must be singular.

mattclay committed:
Rename `is_install` and `INSTALL_ROOT`.

This makes it more obvious what the variables are referencing and keeps them consistent.

- `is_install` -> `is_ansible`
- `INSTALL_ROOT` -> `ANSIBLE_ROOT`

jborean93 commented on test/runner/lib/util.py:
```diff
@@ -78,6 +78,42 @@
+Text = type(u'')
+
+
+def to_optional_bytes(value, errors='strict'):  # type: (t.Optional[t.AnyStr], str) -> t.Optional[bytes]
```
Is it worthwhile taking a copy of the text utils we already have in Ansible. They have a `nonstring` option to define things like what to do when a None object is being encoded.

mattclay replied:
The optional versions exist so that the PEP 484 type hints are accurate. The optional versions may return `None`, while the non-optional versions do not.
